feat: add Valibot support with `valibot$` to qwik-city

brandonpittman opened the pull request:
# Overview

Adds Valibot support with a `valibot$` function similarly to how Zod is supported by `zod$`.

Fixes #4924 

# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests / types / typos

# Description

Requires: 

- `valibot`

# Use cases and why

So you can use Valibot to validate your actions

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

gioboa commented:
Thanks for your nice PR, can you add the documentation for this please?

brandonpittman commented:
> Thanks for your nice PR, can you add the documentation for this please?

Not done, yet. Will update docs tomorrow!

fabian-hiller commented:
@brandonpittman yes, this should work via the `.schema` key: https://github.com/fabian-hiller/valibot/blob/v0.13.1/library/src/schemas/object/object.ts#L65

fabian-hiller commented:
Note that you must also include `BaseSchemaAsync` and `ObjectShapeAsync` to support async validation.

brandonpittman commented:
> Note that you must also include `BaseSchemaAsync` and `ObjectShapeAsync` to support async validation.

@fabian-hiller 

It's not clear if we need async support. I've updated my PR with working code that satisfies TypeScript. I've removed any async types though. Can you take a look, please?

brandonpittman commented on packages/qwik-city/runtime/src/server-functions.ts:
```diff
@@ -230,6 +231,58 @@ export const validatorQrl = ((
+          return {
+            success: false,
+            status: 400,
+            error: v.flatten(result.issues),
```
Should I return a `fieldErrors` key to match the existing Zod implementation?

brandonpittman replied:
I’d rather use Valibot’s return value as is, but I wanted to run it by someone just to check. Thanks.

fabian-hiller commented:
> It's not clear if we need async support.

Since the validation is done in an async function, we could enable async validation with `safeParseAsync`, `BaseSchemaAsync` and `ObjectShapeAsync`. This would allow, for example, to do network requests as part of the validation. If someone uses only sync validation, this is not a disadvantage. async is compatible with sync, but sync is not compatible with async.

fabian-hiller commented:
If you want to add async support, you need to replace `safeParse` with `safePareseAsync` and also add `BaseSchemaAsync` and `ObjectShapeAsync` to the types.

brandonpittman commented:
> If you want to add async support, you need to replace `safeParse` with `safePareseAsync` and also add `BaseSchemaAsync` and `ObjectShapeAsync` to the types.

Okay. Will update it. Thanks.

fabian-hiller commented:
You must not remove `BaseSchema` and `ObjectShape` from the types. Otherwise the sync validation will not be accepted.

brandonpittman commented:
> You must not remove `BaseSchema` and `ObjectShape` from the types. Otherwise the sync validation will not be accepted.

@fabian-hiller 

Should I use BaseSchema instead of ObjectSchema<ObjectShape>?

mhevery commented:
Thanks for working on this! ❤️

fabian-hiller commented:
> Should I use `BaseSchema` instead of `ObjectSchema<ObjectShape>`? When I tried `BaseSchema` the `.schema === "object'` check encountered a type error.

I think I would write `BaseSchema | BaseSchemaAsync | ObjectShape | ObjectShapeAsync`. If we recognize that an object with `typeof obj._parse === "function"` is passed, it is a scheme that can be passed directly in `safeParseAsync`. Otherwise the object must be passed into `objectAsync` before we call `safeParseAsync`.

> Do I need to do something like `obj.async ? safeParseAsync(obj, data) : safeParse(obj, data)` or can I just give `safeParseAsync` a synchronous schema?

This is not required. `safeParseAsync` can also handle `BaseSchema`.

zanettin commented:
Thank you very much @brandonpittman and @fabian-hiller 🙏 

One thought from my side regarding exporting both libs now from the `qwik-city` pkg is, if we should think about flag the `zod$` and `z` exports as deprecated? it shouldn't be our long term goal to keep both within the framework imo. end users would still be able to additionally add `zod$` if they want to keep it. @mhevery would like to pass that to you and the team 😄

mhevery commented on packages/qwik-city/runtime/src/index.ts:
```diff
@@ -58,10 +58,12 @@ export { routeAction$, routeActionQrl } from './server-functions';
 
 export { z } from 'zod';
+import * as v from 'valibot';
+export { v };
```
I don't think we should be exporting `valibot` like this. (I know we do that with `z` but `z` is a specific object whereas `v` is a module. I don't think we should export anything, and people can directly import from `valibot` to get the right things.

brandonpittman replied:
Worried about the same thing myself. Removed the re-export.

mhevery commented on packages/docs/src/routes/docs/(qwikcity)/action/index.mdx:
```diff
@@ -216,6 +216,11 @@ export const useAddUser = routeAction$(
     firstName: z.string(),
     lastName: z.string(),
   })
+  // Or, with Valibot:
```
These comments will become a liability over time as we change things and we can't verify that they actually work.

Can we change them to an actual working example? (I know I am creating more work, please?) See https://github.com/BuilderIO/qwik/blob/main/packages/docs/src/routes/docs/integrations/modular-forms/index.mdx#L165-L231 for how it can be done

1. Create a working example here: https://github.com/BuilderIO/qwik/blob/main/packages/docs/src/routes/demo/integration/modular-forms/index.tsx
2. insert `<CodeSandbox src="/src/routes/demo/integration/modular-forms/index.tsx">`
3. Run `codesandbox.sync` in `packages/docs` to sync the `CodeSandbox` with the actual example.

brandonpittman replied:
@mhevery 

I added CodeSandbox demos for the simple and advanced Valibot sections. Do we need CodeSandboxes for the smaller third and fourth sections?

mhevery commented on packages/docs/src/routes/docs/(qwikcity)/action/index.mdx:
```diff
@@ -275,6 +281,20 @@ export const useAddUser = routeAction$(
+          ? v.optional(v.string())
+          : v.string([v.minLength(1)])
+
+      return z.object({
```
Perfect example right here. This comment is wrong as we are mixing both valibot and zod. Imagine the frustration of the person trying to use this. We need real working code not comments.

brandonpittman replied:
Added separate sections for Valibot.

brandonpittman commented on packages/qwik-city/runtime/src/api.md:
```diff
@@ -441,6 +442,16 @@ export const useLocation: () => RouteLocation;
+export { v }
+
+// @public (undocumented)
+export const valibot$: any;
```
I also thought this was a problem, but it's how `zod$` is currently. Figured there must be a reason for it.

https://github.com/BuilderIO/qwik/blob/7a5a3814dcf20e6def262a8065484c6cecc06830/packages/qwik-city/runtime/src/api.md?plain=1#L467-L480

mhevery replied:
hmm... that is a question for @manucorporat

brandonpittman commented on packages/qwik-city/runtime/src/server-functions.ts:
```diff
@@ -230,6 +231,60 @@ export const validatorQrl = ((
+/**
+ * @public
+ */
+export const valibot$ = /*#__PURE__*/ implicit$FirstArg(valibotQrl) as any;
```
I was following `zod$`'s lead.

https://github.com/BuilderIO/qwik/blob/7a5a3814dcf20e6def262a8065484c6cecc06830/packages/qwik-city/runtime/src/server-functions.ts#L333-L336

brandonpittman commented on packages/qwik-city/runtime/src/server-functions.ts:
```diff
@@ -41,6 +41,7 @@ import type {
 } from './types';
 import { useAction, useLocation, useQwikCityEnv } from './use-functions';
 import { z } from 'zod';
+import * as v from 'valibot';
```
Fixed.

brandonpittman commented:
<img width="257" alt="CleanShot 2023-08-30 at 10 18 03@2x" src="https://github.com/BuilderIO/qwik/assets/967145/1a398909-97f9-4927-9da7-78d6ee1aedeb">

@mhevery 

I updated the validation heading structure because I added separate Valibot sections. Is this okay?

brandonpittman commented:
No problems with the runtime code, but there's a lot of Zod-specific types mixed into the routeAction-adjacent types (starting here: https://github.com/BuilderIO/qwik/blob/d27177d1e4749805492d23164f9e655142f24601/packages/qwik-city/runtime/src/types.ts#L810-L813C2). A little nervous about mucking around with them.

brandonpittman commented:
> Thank you @brandonpittman for your efforts! 💙

@fabian-hiller I think the last thing should be adjusting the action value type. After removing the any, it’s not correctly type hinting the error key when failed is true. I still need to figure that out.

fabian-hiller commented:
Can you highlight what you mean? Then I can have a look at it.

brandonpittman commented:
@fabian-hiller 

starts here: https://github.com/BuilderIO/qwik/blob/3d4715e4a4b76eed3db3163f7850bee8d7dfc089/packages/qwik-city/runtime/src/types.ts#L827-L851

continues through to here: https://github.com/BuilderIO/qwik/blob/3d4715e4a4b76eed3db3163f7850bee8d7dfc089/packages/qwik-city/runtime/src/types.ts#L419-L608

fabian-hiller commented:
Thanks for the info. I will have a closer look next week.

brandonpittman commented:
> Thanks for the info. I will have a closer look next week.

Hopefully I can figure it out before then. 😆

fabian-hiller commented:
Ok. Sorry. I'm moving this weekend. Otherwise I would have looked at it right away.

brandonpittman commented:
> Ok. Sorry. I'm moving this weekend. Otherwise I would have looked at it right away.

@fabian-hiller Sorry, I hope that didn't sound like I was rushing you. I just meant I really want to solve the problem. I'm very grateful for all the help you've given me so far. 🙇🏻

fabian-hiller commented:
I will try to review the changes in the next few days.

fabian-hiller commented on packages/docs/src/routes/demo/valibot/advanced/index.tsx:
```diff
@@ -0,0 +1,64 @@
+
+export const useAddUser = routeAction$(
+  async (user) => {
+    // The "user" is still strongly typed, but firstname
```
Change "firstname" to "first name"

fabian-hiller reviewed (changes requested):
Maybe for example @mhevery or @zanettin can leave a 👍 or 👎 for confirmation on my comments. Once that's resolved, I'll check the actual implementation.

brandonpittman commented on packages/docs/package.json:
```diff
@@ -74,5 +74,8 @@
     "start": "pnpm dev"
+  },
+  "dependencies": {
+    "valibot": "^0.13.0"
```
Done

mhevery replied:
the vilbot demo is part of docs, so. think `devDependencies` would be wrong. Correct as is..

fabian-hiller replied:
Yes, I thought so too. However, then quite a few dependencies are mapped incorrectly. It doesn't make sense that the docs currently only have dev dependencies and no normal dependencies. Or is there a reason for this?

wmertens replied:
Hmm unless valibot is especially kept out of the bundle for some reason, it should be a devDependency. And I don't see a vite config that keeps it out.

fabian-hiller replied:
Do you know why the docs only have dev dependencies and no normal dependencies?

mhevery commented on packages/docs/src/routes/api/qwik-city/index.md:
````diff
@@ -641,6 +641,31 @@ useNavigate: () => RouteNavigate;
 
+## valibot$
+
+```typescript
````
Can we turn this into a real interactive example? Add `<CodeSandbox>` example: https://github.com/BuilderIO/qwik/edit/main/packages/docs/src/routes/docs/(qwik)/getting-started/index.mdx

mhevery commented on packages/docs/src/routes/docs/(qwikcity)/action/index.mdx:
```diff
@@ -148,41 +150,30 @@ export interface EditFormProps {
-const EditForm = component$(
-  ({ item, editingIdSignal }: EditFormProps) => {
-    const editAction = useEditFromListAction();
+const EditForm = component$(({ item, editingIdSignal }: EditFormProps) => {
```
Should this be `<Codesandbox>`?

mhevery reviewed (commented):
Love the direction. Let me know when it is ready for in depth review.

fabian-hiller commented:
Thank you @brandonpittman. I will check your changes tomorrow.

brandonpittman commented:
@fabian-hiller 

Is there any way to get type hints out of `flatten()`? `FlatErrors` isn't generic, so I can't see a way to pass a schema along to get type hints from the return value of `flatten()`.

fabian-hiller commented:
Currently this is not possible. Do you know if this works with Zod? I think it will be difficult to make it completely type-safe.

brandonpittman commented:
> Currently this is not possible. Do you know if this works with Zod? I think it will be difficult to make it completely type-safe.

@fabian-hiller 

Yeah, the current `zod$` function uses [Zod's `inferFlattenedErrors`](https://github.com/colinhacks/zod/blob/f59be093ec21430d9f32bbcb628d7e39116adf34/src/ZodError.ts#L7-L16).

brandonpittman commented:
I'm unsure what to do here. The existing types have Zod-specific stuff everywhere and I'm everything I try clashes with the Zod stuff. I'd love it if @manucorporat could look at this since he wrote the original types.

mhevery commented:
> I'm not sure about what to do with the types. The existing types have Zod-specific stuff everywhere and everything I try to get correct type-hinting for Valibot clashes with the Zod stuff. I'd greatly appreciate it if @manucorporat could look at this since he wrote the original types.

maybe I can help if you can point me to the issue...

fabian-hiller commented:
I will also be reviewing the code thoroughly over the weekend. This issue has the highest priority now. I would like to have it merged before my next big annoucnment next week or the week after.

brandonpittman commented:
@mhevery 

1. The first problem I'm having is in `packages/qwik-city/runtime/src/types.ts`. The return type from `valibot$` seems to be returning the same thing that `zod$` does. There are a bunch of nested types for the action constructor and I never could figure out how to separate the Zod types so that I could have Valibot types be accepted as well.
2. The second problem is that Valibot's `FlatErrors` type can't take a generic. Even if I could get `valibot$` to type hint the return so that `flatten()`'s `nested` key was hinted, the contents of `nested` can't be passed along.

mhevery commented:
I will have a look, sorry it is taking me so long to get to this...

fabian-hiller commented:
I won't get to my review until next weekend. Last weekend I did not find time for it. 😕

fabian-hiller commented:
> Zod has a couple helpers for inferring flattened error types.

Thank you for the hint. I will add such a type to Valibot too.

fabian-hiller commented:
@brandonpittman do you prefer that I edit the PR directly (if I get permission to do so) or add comments to your code via GitHub review instead?

brandonpittman commented:
> @brandonpittman do you prefer that I edit the PR directly (if I get permission to do so) or add comments to your code via GitHub review instead?

@fabian-hiller Please edit the PR. That’s always easier than parsing comments.

fabian-hiller commented:
> Thank you for the hint. I will add such a type to Valibot too.

The next version of Valibot will bring improved types for `flatten` and `FlatErrors`.

fabian-hiller commented:
Valibot v0.18.0 is now available. You can type `flatten` and `FlatErrors` using a generic similar to Zod.

mhevery commented:
> 1. The first problem I'm having is in `packages/qwik-city/runtime/src/types.ts`. The return type from `valibot$` seems to be returning the same thing that `zod$` does. There are a bunch of nested types for the action constructor and I never could figure out how to separate the Zod types so that I could have Valibot types be accepted as well.

So I tried to build it locally but the demo files are failing to build. Something is off from your types...

Not sure how you are building it but here is my process

1. open `packages/docs/package.json`
2. change to `"@builder.io/qwik-city": "wqrkspace:*",`
3. from the project root run `npm build`
4. Notice build errors is `packages/docs/src/routes/demo/valibot/advanced/index.tsx` and `packages/docs/src/routes/demo/valibot/simple/index.tsx`


What kind of help do you need from me to move this forward?

fabian-hiller commented:
I'll probably finally get around to thoroughly reviewing this PR this weekend. I will then commit directly to the PR and make improvements. I hope that this is possible without special permissions, since I am not an official maintainer of the Qwik repo.

fabian-hiller commented:
@brandonpittman a update for you. I have not forgotten about this PR and will finish it as soon as possible. We are currently working on a major structural change to Valibot. As soon as this is done, I will working on this PR.

wmertens commented:
So the current state is that @fabian-hiller will make changes when he finds the time, right?

fabian-hiller commented:
Yes, this is the current state. But everyone is welcome to continue working on it before I find the time. In about 3 weeks I will be able to fully focus on my open source work again and will probably finish the PR before Christmas.

fabian-hiller commented:
I plan to work on and finalize this PR by the end of next week.

fabian-hiller commented:
I am very sorry that this is taking so long. I expect to finalize this PR next week.

shairez commented:
Thanks @fabian-hiller ! can't wait!

fabian-hiller commented:
Other things came up that I wanted to preordain. This PR is on my list for the next two weeks.

fabian-hiller commented:
There is a Valibot PR I want to review and maybe merge before I finish this implementation because some types may change.

fabian-hiller commented:
Short update: I am basically rewriting Valibot at the moment. I am trying to improve the API design, performance, bundle size and types, but this will take some time. I also have to rewrite the documentation. I don't want to make any more promises about the timing of this PR. As soon as I feel ready, I will finish the implementation.

brandonpittman commented:
The funny thing is, the implementation works it’s just the types I wanted help with.

fabian-hiller commented:
I expect a lot of breaking changes for the next release. So even though the implementation of this PR is almost done, it might make sense to wait to not confuse people.

brandonpittman commented:
@fabian-hiller Can you tell me what is going to change with Valibot that might affect this?

I’m gonna take a stab at fixing the types again. (Making the action types less Zod-specific in another PR first might be better though.)

wmertens commented:
@brandonpittman I spoke with him recently, there's a lot of changes to the API coming, as I understand it. So that zod decoupling first send like a good idea

fabian-hiller commented:
The API will be mostly the same (except for the new `pipe` function), but the types will change and lead to a breaking change for the `valibot$` implementation. Therefore, I recommend adding the `valibot$` API to Qwik once I am done with the rewrite and start preparing our v1 release.

brandonpittman commented:
> @brandonpittman Are you still working on this or is it abandoned?

@fabian-hiller kept saying he would work on it, so I let it go. Asked for help from @manucorporat, who wrote the types I was having trouble with, but never got a reply, so I let it go. The JS worked fine. It was just the types I couldn’t get right and couldn’t get any guidance on it. 

Also, Fabian said the types were going to change for Valibot so I didn’t see the point in working on it alone.

brandonpittman commented:
JS works but couldn’t get any assistance getting the types to work.

gioboa commented:
I'm so sorry for that @brandonpittman 
The idea to support Valibot is great btw, thanks for your hard work on this PR.
We appreciate your effort. I'm looking forward to implementing this feature after v2.0

fabian-hiller commented:
I still plan to finish this PR, but unfortunately I had to give other things, like rewriting Valibot and updating our docs, a higher priority. I think Valibot is on a pretty good track to reach v1 RC in August. I also plan to work more with Qwik soon. Therefore, I hope to be able to work on it in the next two months. However, everyone should feel welcome to take over this PR to speed up the process. I am happy to answer questions and review everything.

fabian-hiller commented:
I have reviewed the `zod$` implementation, but to be honest, it is not in good shape. Many types are wrong, typecasting is used, and many edge cases are not considered. Fixing this is hard as I do not have a broader view of everything. Also, it would most likely lead to breaking changes. I may create a new PR to add `valibot$`, but in general I recommend rewriting and fixing the entire implementation at some point.

brandonpittman commented:
> I have reviewed the `zod$` implementation, but to be honest, it is not in good shape. Many types are wrong, typecasting is used, and many edge cases are not considered. Fixing this is hard as I do not have a broader view of everything. Also, it would most likely lead to breaking changes. I may create a new PR to add `valibot$`, but in general I recommend rewriting and fixing the entire implementation at some point.

@fabian-hiller You mean my fixes or the original implementation?

fabian-hiller commented:
The original implementation.
